File: alljoyn/services/notification/java/NotificationService/src/org/alljoyn/ns/NotificationSender.java
/******************************************************************************
 * Copyright AllSeen Alliance. All rights reserved.
 *
 *    Permission to use, copy, modify, and/or distribute this software for any
 *    purpose with or without fee is hereby granted, provided that the above
 *    copyright notice and this permission notice appear in all copies.
 *
 *    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 *    WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 *    MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 *    ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 *    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 *    ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 *    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 ******************************************************************************/

package org.alljoyn.ns;

import java.util.List;
import java.util.Map;
import java.util.UUID;

import org.alljoyn.about.AboutKeys;
import org.alljoyn.ns.commons.GenericLogger;
import org.alljoyn.ns.commons.NativePlatform;
import org.alljoyn.ns.commons.NativePlatformFactory;
import org.alljoyn.ns.commons.NativePlatformFactoryException;
import org.alljoyn.ns.transport.Transport;


/** 
 * The class is used to trigger sending a notification
 */
public class NotificationSender {
	private static final String TAG = "ioe" + NotificationSender.class.getSimpleName();
	
	/**
	 * The TTL low limit in seconds  
	 */
	public static final int MESSAGE_TTL_LL = 30;
	
	/**
	 * The TTL upper limit in seconds
	 */
	public static final int MESSAGE_TTL_UL = 43200;
		
	/**
	 * Reference to native platform object
	 */
	private NativePlatform nativePlatform;

	/**
	 * Constructor
	 * @throws NotificationServiceException
	 */
	public NotificationSender() throws NotificationServiceException {
		try {
			nativePlatform       = NativePlatformFactory.getPlatformObject();
			GenericLogger logger = nativePlatform.getNativeLogger();
			logger.info(TAG,  "Notification Sender created");
		} catch (NativePlatformFactoryException npfe) {
			throw new NotificationServiceException("Failed to create Notification Service: " + npfe.getMessage());
		}
	}

	/**
	 * Send the Notification message
	 * @param notification The notification object to be sent
	 */
	public void send(Notification notification, int ttl) throws NotificationServiceException {
		GenericLogger logger = nativePlatform.getNativeLogger();
		
		if ( ttl < MESSAGE_TTL_LL || ttl > MESSAGE_TTL_UL ) {
			logger.error(TAG, "The allowed TTL range is between '" + MESSAGE_TTL_LL + "' and '" + MESSAGE_TTL_UL + "'");
			throw new NotificationServiceException("The allowed TTL range is between '" + MESSAGE_TTL_LL + "' and '" + MESSAGE_TTL_UL + "'");
		}
		
		Transport transport = Transport.getInstance();
		
		//get a map of the PropertyStore properties 
   	    Map<String, Object> props = transport.readAllProperties();
   	    
   	    UUID appId = transport.getAppId(props);
   	    
   	    String deviceId   = (String)props.get(AboutKeys.ABOUT_DEVICE_ID);
   	    if ( deviceId == null || deviceId.length() == 0 ) {
   	    	logger.error(TAG, "The DeviceId is NULL or empty");
   	    	throw new NotificationServiceException("The DeviceId is not set in the PropertyStore");
   	    }
   	    
   	    String deviceName = (String)props.get(AboutKeys.ABOUT_DEVICE_NAME);
   	    if ( deviceName == null || deviceName.length() == 0 ) {
   	    	logger.error(TAG, "The DeviceName is NULL or empty");
   	    	throw new NotificationServiceException("The DeviceName is not set in the PropertyStore");
   	    }
   	    
   	    String appName    = (String)props.get(AboutKeys.ABOUT_APP_NAME); 
   	    if ( appName == null || appName.length() == 0 ) {
   	    	logger.error(TAG, "The AppName is NULL or empty");
   	    	throw new NotificationServiceException("The AppName is not set in the PropertyStore");
   	    }
   	    
		NotificationMessageType messageType      = notification.getMessageType();
		List<NotificationText> text              = notification.getText();
		List<RichAudioUrl> richAudioUrl          = notification.getRichAudioUrl();
		Map<String, String> customAttributes     = notification.getCustomAttributes();
		String richIconUrl                       = notification.getRichIconUrl();
		String richIconObjPath                   = notification.getRichIconObjPath();
		String richAudioObjPath                  = notification.getRichAudioObjPath();
		String responseObjectPath                = notification.getResponseObjectPath();
		
		String audioUrl = null;
		if (richAudioUrl != null && richAudioUrl.size() > 0) {
			audioUrl = richAudioUrl.get(0).getUrl();
		}		
		
		logger.debug(TAG, "Sending message with message type: " + messageType + ", TTL: " + ttl + ", message: " + text.get(0).getText() + ", richIconUrl: '" + richIconUrl + "' RichAudioUrl: '" + audioUrl + ", richIconObjPath: '" + richIconObjPath + "' RichAudioObjPath: '" + richAudioObjPath + "'");
		PayloadAdapter.sendPayload(deviceId, deviceName, appId, appName, messageType, text, customAttributes, ttl, richIconUrl, richAudioUrl, richIconObjPath, richAudioObjPath, responseObjectPath);
	}//send

	/**
	 * Delete the last message by a given message type
	 * @param messageType
	 * @throws NotificationServiceException if message type was not defined
	 */	
	public void deleteLastMsg(NotificationMessageType messageType) throws NotificationServiceException  {
		if(messageType == null ) {
			throw new NotificationServiceException("MessageType parameter must be set");
		}
		Transport.getInstance().deleteLastMessage(messageType);
	}//deleteLastMsg

	

}//NotificationSenderImpl

File: services/base/notification/cpp/src/NotificationSender.cc
/******************************************************************************
 * Copyright AllSeen Alliance. All rights reserved.
 *
 *    Permission to use, copy, modify, and/or distribute this software for any
 *    purpose with or without fee is hereby granted, provided that the above
 *    copyright notice and this permission notice appear in all copies.
 *
 *    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 *    WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 *    MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 *    ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 *    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 *    ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 *    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 ******************************************************************************/

#include <alljoyn/notification/NotificationSender.h>
#include <alljoyn/notification/NotificationService.h>
#include <alljoyn/notification/Notification.h>

#include "PayloadAdapter.h"
#include "Transport.h"
#include "NotificationConstants.h"
#include <string.h>
#include <alljoyn/notification/LogModule.h>

using namespace ajn;
using namespace services;
using namespace nsConsts;
using namespace qcc;

NotificationSender::NotificationSender(AboutData* aboutdata) :
    m_aboutdata(aboutdata), m_PropertyStore(NULL)
{

}

NotificationSender::NotificationSender(PropertyStore* propertyStore) :
    m_aboutdata(NULL), m_PropertyStore(propertyStore)
{

}

QStatus NotificationSender::send(Notification const& notification, uint16_t ttl)
{
    QCC_DbgTrace(("Send Message called"));

    //Validations
    if (notification.getMessageType() < 0 || notification.getMessageType() >= MESSAGE_TYPE_CNT) {
        QCC_LogError(ER_BAD_ARG_1, ("MessageType sent is not a valid MessageType"));
        return ER_BAD_ARG_1;
    }

    if (notification.getText().size() == 0) {
        QCC_LogError(ER_BAD_ARG_1, ("There must be at least one notification defined"));
        return ER_BAD_ARG_1;
    } else if ((TTL_MIN > ttl) || (ttl  > TTL_MAX)) {   // ttl value is not in range
        QCC_LogError(ER_BAD_ARG_2, ("TTL sent is not a valid TTL value"));
        return ER_BAD_ARG_2;
    }

    ajn::BusAttachment* pBus = Transport::getInstance()->getBusAttachment();
    String originalSender = pBus ? pBus->GetUniqueName() : "";
    if (m_aboutdata) {
        return PayloadAdapter::sendPayload(m_aboutdata,
                                           notification.getMessageType(), notification.getText(),
                                           notification.getCustomAttributes(), ttl, notification.getRichIconUrl(),
                                           notification.getRichAudioUrl(), notification.getRichIconObjectPath(),
                                           notification.getRichAudioObjectPath(), notification.getControlPanelServiceObjectPath(),
                                           originalSender.c_str());
    }

    if (m_PropertyStore) {
        return PayloadAdapter::sendPayload(m_PropertyStore,
                                           notification.getMessageType(), notification.getText(),
                                           notification.getCustomAttributes(), ttl, notification.getRichIconUrl(),
                                           notification.getRichAudioUrl(), notification.getRichIconObjectPath(),
                                           notification.getRichAudioObjectPath(), notification.getControlPanelServiceObjectPath(),
                                           originalSender.c_str());
    }
    return ER_FAIL;
}

QStatus NotificationSender::deleteLastMsg(NotificationMessageType messageType)
{
    QCC_DbgTrace(("Delete Last Message called"));

    //Validation
    if (messageType < 0 || messageType >= MESSAGE_TYPE_CNT) {
        QCC_LogError(ER_BAD_ARG_1, ("MessageType sent is not a valid MessageType"));
        return ER_BAD_ARG_1;
    }

    Transport* transport = Transport::getInstance();
    return transport->deleteLastMsg(messageType);
}

